Add polyvar docs (latest) (#173)
reason-association/rescript-lang.org: pages/docs/manual/latest/polymorphic-variant.mdx

---
title: "Polymorphic Variant"
description: "The Polymorphic Variant data structure in ReScript"
canonical: "/docs/manual/latest/polymorphic-variant"
---

# Polymorphic Variant

Now that we know what [variant types](./variant) are, let's dive into a more specific version, so called polymorphic variants (or poly variants).

First off, here are some key features:

> - Poly variants are **structurally typed** (in comparison to **nominally typed** variants). They can be used without an explicit type definition.

Reviewer: Was also thinking about this. Do we have some explanations on rescript org we can link too?

> - Poly variants are **structurally typed** (in comparison to **nominally typed** variants). They can be used without an explicit type definition.

Author: maybe it's better to explain the term "nominally typed" in the variants section, and later on link back to it? not sure where to put our definition of structurally typed.. down below we discuss the structural attributes anyways?

> - Poly variants are **structurally typed** (in comparison to **nominally typed** variants). They can be used without an explicit type definition.

Author: I tried to explain it in a few words instead. I'd like to prevent wikipedia links, since they invite to go into an unnecessary rabbit hole

> - They allow easier JavaScript interop (compile to strings / objects with predictable `NAME` and `VAL` attribute) and don't need explicit runtime conversions, unlike common variants.

Reviewer: It would be great if we had some practical use cases/examples:
```rescript
@module("fs")
external openSync: (
  path,
  @bs.string
  [
    | @bs.as("r") #Read
    | @bs.as("r+") #Read_write
    | @bs.as("rs+") #Read_write_sync
    | @bs.as("w") #Write
    | @bs.as("wx") #Write_fail_if_exists
    | @bs.as("w+") #Write_read
    | @bs.as("wx+") #Write_read_fail_if_exists
    | @bs.as("a") #Append
    | @bs.as("ax") #Append_fail_if_exists
    | @bs.as("a+") #Append_read
    | @bs.as("ax+") #Append_read_fail_if_exists
  ],
) => unit = "openSync" 
```
```

Reviewer: Maybe as example in the binding importance section below.

Reviewer 2: why do you need `@as` since you can use #\"r+" directly

Author: added an example that is more fitting and uses escape syntax

> - Due their structural nature, they oftentimes cause tricky type checking errors when types don't match up, which makes them a more advanced feature.

Reviewer 3: ```suggestion
- Due to their structural nature, they oftentimes cause tricky type checking errors when types don't match up, which makes them a more advanced feature.
```

Reviewer 2: it is the combination of `structural + recursive types ` which makes the error message confusing

Author: I mentioned that somewhere in the variant / poly variant comparison in the end. Couldn't replicate any example where an error message gets completely out of hand, unless you are using very thorough libraries such as Tyxml

## Basics

This is how you'd define a poly variant type with an exact set of constructors:

```res
// Note the surrounding square brackets, and # for constructors
type color = [ #Red | #Green | #Blue ]
```

Here is how you'd construct a poly variant value:

```res

> // This doesn't actually need any color type definition

Reviewer 3: Maybe color could be something different than `Red` (like `Black`) in relation to use a value that is not in the type definition above.

Reviewer: Yes, same thought. Or switch the paragraphs. It's a bit strange that we first have a type declaration and then say we don't need it.

Author: Switched paragraphs!

// beforehand

> let myColor = #Red

Reviewer: Can we add the js output?

```

> We can also use poly variant types in annotations without an explicit type definition:

Reviewer 3: ```suggestion
We can also inline poly variant type definitions in annotations without an explicit type definition:
```

```res
let render = (color: [#Red | #Green | #Blue]) => {
  switch(color) {
    | _ => Js.log("...")
  }
}

let color: [#Red] = #Red
```

### Constructor Names

Poly variant constructor names are less restrictive than in common variants (e.g. they don't need to be capitalized):

```res
type users = [ #admin | #moderator | #user ]

let admin = #admin
```

In rare cases (mostly for JS interop reasons), it's also possible to define invalid identifiers, such as hypens or numbers:

```res
type numbers = [#\"1" | #\"2"]

> let one = #\"1"

Reviewer: Not sure if this is a good example, but `#\"r+"` could function as an example. As alternative to @bs.as
```rescript
@module("fs")
external openSync: (
  path,
  @bs.string
  [
    | @bs.as("r") #Read
    | @bs.as("r+") #Read_write
    | @bs.as("rs+") #Read_write_sync
    | @bs.as("w") #Write
    | @bs.as("wx") #Write_fail_if_exists
    | @bs.as("w+") #Write_read
    | @bs.as("wx+") #Write_read_fail_if_exists
    | @bs.as("a") #Append
    | @bs.as("ax") #Append_fail_if_exists
    | @bs.as("a+") #Append_read
    | @bs.as("ax+") #Append_read_fail_if_exists
  ],
) => unit = "openSync" 
```

Reviewer: Maybe as example next to the binding importance below.

Reviewer 2: in the backend, for polymorphic variant specifically, it is okay to use any names, and I think we should encourage people to use them directly without using `as`, that's part of the reason I proposed to make it a bit more native `#"hello"` instead of `#\"hello"`

Author: it's a feature that exists, so it needs to be documented

```

### Constructor Arguments

> This is equivalent to what we've already learned with common variants:

Reviewer 3: Maybe link to relevant section in common variants?

```res
type account = [
  | #Anonymous
  | #Instagram(string)
  | #Facebook(string, int)
]

let acc: account = #Instagram("test")
```

### Annotations with Upper / Lower Bound Constraints

> There's also a way to define an "upper" and "lower" bound constraint for a poly variant type (that's why they are called _Polymorphic Variants_). Here is what it looks like in type annotations:

Reviewer: What is the relation between "upper" and "lower" and polymorphic, I don't understand this very well.

Author: yeah that's confusing, removed that

```res
// Only #Red allowed, no upper / lower bound (= exact)
let basic: [#Red] = #Red

// May contain #Red, or any other value (open variant)
// here, foreground will be an inferred type [> #Red | #Green]
let foreground: [> #Red] = #Green

// The value must be "one of" #Red | #Blue
// Only #Red and #Blue are valid values
let background: [< #Red | #Blue] = #Red
```

Don't worry about the upper / lower bound feature (aka polymorphism) just yet, since this is a very advanced topic that's often not really needed. For the sake of completeness, we mention a few details about it [later on](#lower--upper-bound-constraints).

## Polymorphic Variants are Structurally Typed

As we've already seen in the section above, poly variants are treated a little bit differently than common variants. Most notably, we don't need any explicit type definition to define a value.

> ```res

Reviewer 2: You may mention one virtue of structural types, for package a and b who share the same structural types, they don't need depend on each other. But it is not the case for norminal types

Author: mentioned it in multiple occasions

```

The compiler will automatically infer the `color` binding as a value of type `[> #Red]`, which means `color` will type check with any other poly variant type that defines `#Red` in its constructors.

This means that you can essentially mix and match poly variant values from different sources, as long as all constructors are defined in the final interface. For example:

```res
type rgb = [#Red | #Green | #Blue]

let colors: array<rgb> = [#Red]

// `other` is inferred as a type of array<[> #Green]>
let other = [#Green]

// Because `other` is of type `array<[> Green]>`,
// this will type check even though we didn't define
// `other`to be of type rgb
let all = Belt.Array.concat(colors, other)
```

As you can see in the example above, the type checker doesn't really care that `color` is not annotated as an `array<rgb>` type. As soon as it hits the first constraint (`Belt.Array.concat`), it will try to check if `colors` and `other` unify into one polymorphic type. If there's a mismatch, you will get an error on the `Belt.Array.concat` call.

**That means that it is very easy to get confusing type errors on the wrong locations!**

For instance, if I'd make a typo like this:

```res

> let other = [#GreeN]

Reviewer 3: ```suggestion
let other = [#Green]
```

Author: hehe, this is the actual trypo I tried to demonstrate: `For instance, if I'd make a typo like this:`

Maybe worth adding a comment on top

Reviewer 3: Oops, sorry, I missed the introductory line before. 😅

```js
// direction.js
enum Direction {
  Up = "UP",
  Down = "DOWN",
  Left = "LEFT",
  Right = "RIGHT",
}

export const myDirection = Direction.Up
```

For this particular example, we can use poly variants to design the type for the imported `myDirection` value:

<CodeTab labels={["ReScript", "JS Output"]}>

```res
type direction = [ #UP | #DOWN | #LEFT | #RIGHT ]
@bs.module("./direction.js") external myDirection: direction = "myDirection"
```

```js
var DirectionJs = require("./direction.js");

var myDirection = DirectionJs.myDirection;
```

</CodeTab>

Since we were using poly variants, the JS Output is practically zero-cost and doesn't add any extra code!

## Lower / Upper Bound Constraints

There are a few different ways to define constraints on a poly variant type, such as `[>`, `[<` and `[`. Some of them were briefly mentioned before, so in this section we will quickly explain what this syntax is about.

**Note:** We added this info for educational purposes. In most cases you will not want to use any of this stuff, since it makes your APIs pretty unreadable / hard to use.

### Exact (`[`)

> This is the simplest poly variant definition, and also the most practical one. Like a common variant type, this one defines an exact set of constructors.

Reviewer 2: this is called closed types

```res
type rgb = [ #Red | #Green | #Blue ]

let color: rgb = #Green
```

In the example above, `color` will only allow one of the three constructors that are defined in the `rgb` type. This is usually the way how poly variants should be defined.

In case you want to define a type that is extensible in polymorphic ways (or in other words, subtyping allowed sets of constructors), you'll need to use the lower / upper bound syntax. 

### Lower Bound (`[>`)

A lower bound defines the minimum set of constructors a poly variant type is aware of. It is also considered an "open poly variant type", because it doesn't restrict any additional values.

Here is an example on how to make a minimum set of `basicBlueTones` extensible for a new `colors` type:

```res

> type basicBlueTone<'a> = [> #Blue | #DeepBlue | #Azuro ] as 'a

Reviewer 3: [Azure](https://en.wikipedia.org/wiki/Azure_(color))?

Reviewer: Would an explanation of the `as 'a`  be needed here?

Author: yeah it would, but not sure how to explain it in words without any concrete example (without being overly abstract)

Reviewer 4: The `as 'a` makes a lot more sense once we understand "type quantification," which I think is pretty foreign/arcane to a lot of people. But it might be helpful to allude to some of the **other** ways that types are quantified, and how the `as` keyword plays a role in determining how "type holes" are "filled."

Author: We will have a dedicated section about polymorphism in an advanced section... i think i will go with a note that does some oversimplification for the case at hand, and later on point to more advanced resources when they are ready.

type color = basicBlueTone<[#Blue | #DeepBlue | #Azuro | #Purple]>

let color:  color = #Purple

// This will fail due to missing minimum constructors:
type notWorking = basicBlueTone<[#Purple]>
```

Here, the compiler will enforce the user to define `#Blue | #DeepBlue | #Azuro` as the minimum set of constructors when trying to extend `basicBlueTone<'a>`. 

### Upper Bound (`[<`)

> The upper bound works in the exact opposite way: The extending type may only use constructors that are stated in the lower bound constraint.

Reviewer: ```suggestion
The upper bound works in the exact opposite way: the extending type may only use constructors that are stated in the lower bound constraint.
```

Here another example, but with red colors:

```res
type validRed<'a> = [< #Fire | #Crimson | #Ash] as 'a
type myReds = validRed<[#Ash]>

// This will fail due to unlisted constructor not defined by the lower bound
type notWorking = validRed<[#Purple]>
```

## Tips & Tricks

> - In most scenarios, you should prefer common variants over polymorphic variants, since they offer better error messages and easier to spot errors in your program.

Reviewer: Would it make sense to add a table with the differences between normal and poly variants summarized?

Should there be an example of when it's the wrong use case for poly variants?

Author: i added a separate "Variants vs PolyVariants" section... but i refrained from bringing up examples because it would be too much

> - In most scenarios, you should prefer common variants over polymorphic variants, since they offer better error messages and easier to spot errors in your program.

Reviewer 2: people have different preferences, in my experience, for recursive types, I would go with normal variants. But for simple data types, polymorphic variant is used from time to time

Author: removed the tips & tricks and added a "Variant vs Poly Variant" section instead with a few notes

> - Polymorphic variants are pretty useful for doing zero-cost interop, e.g. when binding to JavaScript string enums, or to bind seemlessly between a tagged union type in TypeScript and ReScript

Reviewer 3: Another use case is when you need to "extend" an existing type with new values, which is briefly mentioned in PR but could be expanded: https://2ality.com/2018/01/polymorphic-variants-reasonml.html#extending-polymorphic-variants.

Extended variants also combines with the ability of pattern matching a whole set of values with an existing type: 

```ocaml
let extended_color_to_int : extended_color -> int = function
  | `RGBA (r,g,b,a) -> 256 + a + b * 6 + g * 36 + r * 216
  | #color as color -> color_to_int color
```

Not sure if you want to include all this though? :)

Author: My supervisors told me to keep it short 😄 

The pattern matching part _might be interesting_ though.

Note to myself, example above looks like this in ReScript:

```rescript
let extendedColorToInt = (x: extendedColor): int => {
  switch x {
  | #RGBA(r, g, b, a) => 256 + a + b * 6 + g * 36 + r * 216
  | #...color as color => color_to_int
  }
}
```

Reviewer: Yes, I was also gonna suggest explaining the spread syntax. It's a special syntax thing.

Reviewer 2: yeah, we should mention spreading syntax which is a nice sugar

Author: added in a small section
